Implement ValidateUsernameUseCase for Username Validation

Prashant-Android opened the pull request:
This PR introduces the ValidateUsernameUseCase to enforce consistent username validation across the app, addressing the following:

Allowable characters: letters (including non-English), numbers, . and _.
Disallowed characters: whitespace and banned special symbols (e.g., !@#$%^&*()).
Error message updated for clarity.

Prashant-Android committed:
- Added `ValidateUsernameUseCase`.
- Updated `LoginViewModel` and `PersonEditViewModel`.
- Added string resource for invalid username error.

Prashant-Android committed:
Replaced regex with explicit validation, allowing non-English letters and disallowing banned characters.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/person/edit/PersonEditViewModel.kt:
```diff
@@ -356,29 +359,7 @@ class PersonEditViewModel(
     }
 
     private fun validateUsername(username: String): Boolean {
```
This function can be deleted and the validate username use case can be called directly. This function serves no purpose.

mikedawson commented:
This validation needs to be used by PersonAccountEditViewModel - which is where the admin would go to create an account.

Prashant-Android committed:
Use `ValidateUsernameUseCase` in `PersonAccountEditViewModel` and remove username validation from `PersonEditViewModel`.

Prashant-Android committed:
Added username in PersonEditViewModelTest.kt

Prashant-Android committed:
- Fixed PersonEditViewModelTest to expect invalid_username error
- Fixed LoginViewModelTest to handle empty username validation

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/domain/validateusername/ValidateUsernameUseCase.kt:
```diff
@@ -0,0 +1,30 @@
+    operator fun invoke(username: String): String? {
+        val trimmed = username.trim()
+
+        if (trimmed.length !in 3..15) {
```
Clear violation of coding style rule: "never hardcode literals" see https://github.com/UstadMobile/UstadMobile/blob/primary/CODING-STYLE.md#never-hardcode-any-literal-values

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/domain/validateusername/ValidateUsernameUseCase.kt:
```diff
@@ -0,0 +1,30 @@
+     * - Must be 3–15 characters long after trimming
+     *
+     * @param username The username string to validate
+     * @return The valid username (trimmed) if valid, or null if invalid
```
If banned characters are found, then this should remove/replace them with and return whatever is valid 

add an argument: ```invalidReplacement: String = ""```

Any banned characters should then be replaced with the replacement. This is going to be required when running the migration to fix any invalid usernames found.

Prashant-Android committed:
- Add database migration (202->203) to fix existing usernames
- Update username fields to validate input in real-time
- test-end-to-end and testserver-controller folder copied from branch test-validate-username that passed all test cases.

mikedawson commented on lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/ext/UmAppDatabaseMigrations.kt:
```diff
@@ -1653,6 +1653,43 @@ val MIGRATION_201_202 = DoorMigrationStatementList(201, 202) { db ->
     }
 }
 
+val MIGRATION_202_203_SERVER = DoorMigrationStatementList(202, 203) { db ->
```
The migration needs to use the ValidateUsernameUseCase so it follows the don't repeat yourself principle -e.g. the logic for what is, and is not, a valid username, should be in one place.

If you put this migration into ```jvmMain``` source set, you can then use ```db.connection``` to get the underlying JDBC database connection so you can run any JDBC query or statement.

It should go through all usernames in batches, check the username, and then if a username is found where the return result of validateusernameusecase is not the same as the username (e.g. something in the username is invalid), it should be updated (e.g. the update statement should only run on the usernames that need changed).

The migration as-below has different logic. It would not find white space other than space (e.g. won't find tab, carriage return, etc). It also won't find or fix usernames with special characters.

mikedawson replied:
Note: you may need to put the server side migration in app-ktor-server or core/jvmMain source set, otherwise, it won't have access to the ValidateUsernameUseCase.

Prashant-Android committed:
Added 002_010_username_validation testcase

poojaustad committed:
username validation

Prashant-Android committed:
Fixed Validation.

Prashant-Android committed:
Merge remote-tracking branch 'origin/dev-validate-username' into dev-validate-username

# Conflicts:
#	test-end-to-end/android-maestro/e2e-tests/002_010_username_validation.yaml

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/domain/validateusername/ValidateUsernameUseCase.kt:
```diff
@@ -0,0 +1,69 @@
+        return if (containsInvalidChars) {
+            lowercased.map { char ->
+                when {
+                    !isCharacterAllowed(char) -> invalidReplacement
```
What's going on here is two separate logic tracks - one dealing with white space, and another other non-allowed characters? Why not just have isCharacterAllowed return false when character is white space? Why have different logic handling white space vs. other non-allowed characters?

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/login/LoginViewModel.kt:
```diff
@@ -156,9 +159,26 @@ class LoginViewModel(
+        val lastChar = newValue.lastOrNull()
+        val isFirstChar = newValue.length == 1
+
+        val isValid = when {
```
This is repeating the same logic that is in the use case. That's against [coding style](https://github.com/UstadMobile/UstadMobile/blob/primary/CODING-STYLE.md#dont-repeat-yourself) - "Don't repeat yourself" -important principle in software development ( see also https://en.wikipedia.org/wiki/Don%27t_repeat_yourself ).

There is now separate logic validating the username in three places now: PersonAccountEdit, Login, and the ValidateUsernameUseCase. That is very problematic : for all reasons mentioned in the Wikipedia article e.g. more likely to have a bug, if something needs changed it needs changed in multiple places. Same concept as I referenced in Meshrabiya for converting ip addresses to Int etc.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/login/LoginViewModel.kt:
```diff
@@ -156,9 +159,26 @@ class LoginViewModel(
-            prev.copy(username = username)
+    fun onUsernameChanged(newValue: String) {
+        // If no input or input is valid (including whitespace)
+        val lastChar = newValue.lastOrNull()
```
This logic fails when anytime the user has changed any character which is not at the end of the text field.

Prashant-Android committed:
Logic correction in ValidateUsernameUseCase

Prashant-Android committed:
Add ValidateUsernameUseCase with validation result enums and integrate username validation in final submit button

mikedawson commented:
Closing as this is now replaced with #999
